server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalDateHistogram.java

```diff
@@ -444,26 +444,22 @@ private void addEmptyBuckets(List<Bucket> list, ReduceContext reduceContext) {
     @Override
     public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
         List<Bucket> reducedBuckets = reduceBuckets(aggregations, reduceContext);
-
-        // adding empty buckets if needed
-        if (minDocCount == 0) {
-            addEmptyBuckets(reducedBuckets, reduceContext);
-        }
-
-        if (InternalOrder.isKeyAsc(order) || reduceContext.isFinalReduce() == false) {
-            // nothing to do, data are already sorted since shards return
-            // sorted buckets and the merge-sort performed by reduceBuckets
-            // maintains order
-        } else if (InternalOrder.isKeyDesc(order)) {
-            // we just need to reverse here...
-            List<Bucket> reverse = new ArrayList<>(reducedBuckets);
-            Collections.reverse(reverse);
-            reducedBuckets = reverse;
-        } else {
-            // sorted by compound order or sub-aggregation, need to fall back to a costly n*log(n) sort
-            CollectionUtil.introSort(reducedBuckets, order.comparator(null));
+        if (reduceContext.isFinalReduce()) {
+            if (minDocCount == 0) {
+                addEmptyBuckets(reducedBuckets, reduceContext);
+            }
+            if (InternalOrder.isKeyDesc(order)) {
```

Comment:
It took me a few minutes to convince myself that this works.  :)  

On first glance the behavior changes because sorting is only done on the final reduction now (instead of every reduction), which I thought might break `reduceBuckets()` as that relies on consistent ordering.  But histos/date_histos sort their shard results by `key:asc`, so this isn't actually a problem.

Could we add a test to `DateHistogramAggregatorTests`/`HistogramAggregatorTests` that randomly chooses a sort order + `min_doc_count: 0` to ensure this internal contract never changes in the future?  It looks like none of those tests set an order so we might not notice otherwise.

Reply:
I don't think I changed anything around sorting. That if was hard to read, I rewrote but the only actual change is that empty buckets are only added in the final reduction phase. I am working on tests, I will add what you suggest, I had other additions as well in mind but I hit some roadblocks (see #36004). Basically empty buckets were not tested in our unit tests.

Reply:
We chatted about this in slack, and I'm just bad at boolean logic.  All good here :)

Reply:
I added a check to the base class which verifies that no buckets are added in non final reduce phases. Now that we test adding empty buckets for histogram and date histogram aggs, this check makes some sense I think. These are the only aggs that add buckets as part of reduce right? I wonder if I need to check whether there is some other missing test coverage somewhere. 

Also, I worked a bit on increasing test coverage in DateHistogramAggregatorTests like you suggested, and I think I prefer doing it in a follow-up if still necessary.